src/server/pfs/s3/object.go

```diff
@@ -0,0 +1,172 @@
+package s3
+
+import (
+	"bytes"
+	"io"
+	"net/http"
+	"strings"
+	"fmt"
+	"crypto/md5"
+	"encoding/base64"
+
+	"github.com/gogo/protobuf/types"
+	"github.com/pachyderm/pachyderm/src/client"
+)
+
+type objectHandler struct {
+	pc *client.APIClient
+}
+
+func newObjectHandler(pc *client.APIClient) *objectHandler {
+	return &objectHandler{pc: pc}
+}
+
+func (h *objectHandler) get(w http.ResponseWriter, r *http.Request) {
+	repo, branch, file := objectArgs(w, r)
+	branchInfo, err := h.pc.InspectBranch(repo, branch)
+	if err != nil {
+		notFoundError(w, r, err)
+		return
+	}
+	if branchInfo.Head == nil {
+		noSuchKeyError(w, r)
+		return
+	}
+	if isMeta(file) || strings.HasSuffix(file, "/") {
+		invalidFilePathError(w, r)
+		return
+	}
+
+	fileInfo, err := h.pc.InspectFile(branchInfo.Branch.Repo.Name, branchInfo.Head.ID, file)
+	if err != nil {
+		notFoundError(w, r, err)
+		return
+	}
+
+	timestamp, err := types.TimestampFromProto(fileInfo.Committed)
```

Comment:
Note that this timestamp will advance when new files are committed even if this file isn't modified in the commit. This may break the use cases for this, which I assume is doing `If-Modified-Since` requests.

Reply:
That's not great, but this is still better than no timestamp at all. We don't have a way of seeing when a specific file was changed, do we?

Reply:
We do, but it requires a somewhat expensive traversal. `ListFile` with the history flag does this.

Reply:
I just updated to use history - is the new usage correct? Tests are passing, but due to cluster clock drift it's not easy to test for this.

Reply:
I updated this to pull the right timestamp here, but I'm not sure if it's worth it. In addition to the performance overhead, it also means that the objects list different modification times between here vs when you list objects in a bucket. The latter relies on PFS' globbing, which AFAICS cannot be used with history.

Reply:
As per offline discussion, I've reverted this change